apps/api/src/app/shared/interceptors/shared-cache.ts

```diff
@@ -0,0 +1,152 @@
+import { CacheKeyPrefixEnum } from '../services/cache';
+
+export function validateCredentials(keyPrefix: CacheKeyPrefixEnum, credentials: string) {
+  const entitiesEnvironmentLevel = [CacheKeyPrefixEnum.USER, CacheKeyPrefixEnum.INTEGRATION];
+  const splitCredentials = credentials?.split(':').filter((possibleKey) => possibleKey?.length > 0);
+
+  return entitiesEnvironmentLevel.some((cacheKey) => cacheKey === keyPrefix)
+    ? splitCredentials.length === 1
+    : splitCredentials.length === 2;
+}
+
+/**
+ * The data related to the messages stored by the subscriberId
+ * therefore in order to keep the stored data fresh we need to build the key with subscriberId first.
+ * @param key
+ * @param keyConfig
+ */
+export function getIdentifier(key: string, keyConfig: Record<string, unknown>): { key: string; value: string } {
+  const entitiesSubscriberPreferred = [CacheKeyPrefixEnum.MESSAGE_COUNT, CacheKeyPrefixEnum.FEED];
+  const subscriberPreferredKeys = ['subscriberId', '_subscriberId', '_id', 'id'];
+  const idPreferredKeys = ['_id', 'id', '_subscriberId', 'subscriberId'];
+
+  const subscriberPrefKey = subscriberPreferredKeys.find((prefKey) => keyConfig[prefKey]);
+  const subscriberPreferred = { key: subscriberPrefKey, value: keyConfig[subscriberPrefKey] as string };
+
+  const idPrefKey = idPreferredKeys.find((prefKey) => keyConfig[prefKey]);
+  const idPreferred = { key: idPrefKey, value: keyConfig[idPrefKey] as string };
+
+  return entitiesSubscriberPreferred.some((entity) => key.startsWith(entity)) ? subscriberPreferred : idPreferred;
+}
+
+export function getEnvironment(keyConfig: Record<string, unknown>): { key: string; value: string } {
+  return keyConfig._environmentId
+    ? { key: '_environmentId', value: keyConfig._environmentId as string }
+    : keyConfig.environmentId
+    ? { key: 'environmentId', value: keyConfig.environmentId as string }
+    : undefined;
+}
+
+export function buildCredentialsKeyPart(key: string, keyConfig: Record<string, unknown>): string {
+  let credentialsResult = '';
+  const identifier = getIdentifier(key, keyConfig);
+
+  if (identifier?.key) {
+    credentialsResult += ':' + getCredentialWithContext(identifier.key, identifier.value);
+  }
+
+  const environment = getEnvironment(keyConfig);
+
+  if (environment?.key) {
+    credentialsResult += ':' + getCredentialWithContext(environment.key, environment.value);
+  }
+
+  return credentialsResult;
+}
+
+export function getCredentialWithContext(credentialKey: string, credentialValue: string): string {
+  const context = credentialKey.replace('_', '')[0];
+
+  return context + '=' + credentialValue;
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export enum CacheInterceptorTypeEnum {
+  CACHED = 'cached',
+  INVALIDATE = 'invalidate',
+}
+
+export function buildKey(
```

Comment:
I see that this function is too generic and it covers all the cases and builds the different keys for different scenarios.
This one was really hard for me to understand what was going on here.
I think it will keep growing and will be hard to maintain in the future.

So my suggestion is like this, instead of building that key in the `@Cached` and `invalidateCache` functions, we should allow passing the "key string" or the "key builder" function ex.:

```
const keyBuilder = (args:  { _id: string; environmentId: string }) => {
   return `${CacheKeyPrefixEnum.SUBSCRIBER}:${key}=${value}` // ... the implementation 
}

@Cached(CacheKeyPrefixEnum.SUBSCRIBER)

// -- OR --

// the args here are the same args from the function
@Cached((args) => keyBuilder({ _id: args._id,  environmentId: args.environmentId }))
private async getSubscriber({ _id, environmentId }: { _id: string; environmentId: string }) {
...
}
```

Then that key builder function can be reused to invalidate the cache later:

```
invalidateCache({
  service: this.cacheService,
  key: keyBuilder({ _id: command.id, environmentId: command.environmentId }),
});

// -- OR --
@InvalidateCache((args) => keyBuilder({ _id: args._id,  environmentId: args.environmentId }))
```

Example of `Cached` interceptor:
```
export function Cached(key: string | (...args: any[]) => string) {
  ...
  return (target: any, key: string, descriptor: any) => {
    ...
    descriptor.value = async function (...args: any[]) {
      ...
      const cacheKey = typeof key === 'function' ? key(...args) : key ;
      ...
    };
  };
}
```

This way it will be much easier to maintain the key creation logic.

Reply:
Hey @LetItRock I remember that you raised this one before and that is why i created a task regarding it https://linear.app/novu/issue/NV-1304/extract-the-key-creation-logic but i think that tagging in linear is not working so well so a direct ping as well would be a good idea next time.

Regarding the complexity, I tend to agree that it becomes very generic. Do you think that dividing the functionality of buildKey into multiple less generic functions and reusing them would be more maintained and less complex?